Use platform-aware path splitting for interacting with gitignore

deadmoose opened the pull request:
The gitignore library expects path components passed into it, so this was splitting apart on slashes.

On windows, you've got a different path separator, and things get muddy. It's uglier, like the comment says, with there being MULTIPLE potential path separators depending on context (e.g. you're not likely to get `foo/bar\baz` passed back to you from something, but you COULD pass it to os.Open or the like & have it work)

This is what we've been using in-house in our scanning implementation, so I'm surfacing it to see how you'd like to go about incorporating it. I've also got a handful of unit tests for it, but didn't bother bringing those over yet.

It feels a bit wasteful calling filepath.Split over and over to build this slice & reverse it, but it's conceptually straightforward and this doesn't get used on anything big.

I haven't taken the time to fully undesrtand the volume indicator stuff so it's possible this does weird stuff with absolute paths but for this purpose I didn't need to sort it out yet. I _think_ that winds up being the case it's breaking out for an empty file since `/foo/bar/baz` will eventually return `("/", "")` at the root, but I moved on after solving for our immediate need.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved file path handling to ensure compatibility across different operating systems when matching scanning skip patterns.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

deadmoose commented:
And fwiw, some very barebones test:

```
func TestFullySplitPath(t *testing.T) {
  // TODO(dhoover): Make this test more platform agnostic... The whole point is this behaves
  // differently in different places so we cant just embed a unix path.
  // EXCEPT... on windows it actually DOES accept either forward or backslashes as path separators, so :shruggie:

  expected := []string{"foo", "bar", "baz"}

  require.Equal(t, fullySplitPath("foo/bar/baz"), expected)
  require.Equal(t, fullySplitPath("/foo/bar/baz"), expected)
  require.Equal(t, fullySplitPath(filepath.Join(expected...)), expected)
}
```

IIIRC currently github actions here are NOT running the tests on the different target platforms, so that last one which would build it up with backslashes on windows wouldn't ever actually get run in that way.

matiasdaloia commented:
Hi David

LGTM, only thing I would say is that we might want this function inside `internal/utils/file_utils.go`

The other option we have is to use something like 

```go
path := "foo/bar/file.txt"
parts := strings.Split(path, string(os.PathSeparator))
```

However, I think your approach is better because it handles paths independently of current OS.

deadmoose commented:
> The other option we have is to use something like
> 
> ```go
> path := "foo/bar/file.txt"
> parts := strings.Split(path, string(os.PathSeparator))
> ```

Like the comment in there says, it isn't necessarily that simple:

> // We theoretically could split on our platform-native path separator BUT it's not quite that simple: on windows,
> // both forward AND back slashes as path separators, plus there's special handling of potential volume specifiers.

Running through [`filepath.Clean`](https://pkg.go.dev/path/filepath#Clean) would _theoretically_ will normalize it all, but at this point I'm paranoid.

That also ignores the issue around volume specifier. In this case we shouldn't see those since we're only working with relative paths based inside the scanroot and NOT absolute paths, but again I get paranoid about these things getting eventually reused elsewhere without the nuance. E.g. splitting on backslashes would misbehave with things like `C:\foo\bar\baz` or `\\host\share\x`

Depending how interested you are, take a peek at the implementation of `filepath.Split` which just calls `filepathlite.Split`: https://cs.opensource.google/go/go/+/master:src/internal/filepathlite/path.go;l=204-212

That worries about `filepathlite.VolumeName(path)` at the front and then steps with `filepathlite.IsPathSeparator()` as a predicate because it's not a simple check for whether it matches a single known character.

[`path_unix.go`](https://cs.opensource.google/go/go/+/master:src/internal/filepathlite/path_unix.go) is ~50 lines doing the straightforward stuff you'd expect checking slashes, seeing if there's a leading slash, etc.

But then [`path_windows.go`](https://cs.opensource.google/go/go/+/master:src/internal/filepathlite/path_windows.go) is ~325 lines of business logic for every "simple" question.

Again, I'd be open to writing something that uses the same logic as `filepath.Split()` stepping through character by character but cracks the WHOLE thing apart, but sticking to the documented public api like this is probably the easiest way to avoid tripping over any other crazy corner cases the core language folks are hopefully on top of.

matiasdaloia commented:
Yes i agree, your implementation is better than doing the `os.PathSeparator` thing and should be enough. If you can move it to `file_utils` pkg i'll approve and merge it

matiasdaloia commented:
Closing and merged in #108
